### PR TITLE
Speech impediments

### PR DESCRIPTION
**Pull request by Weblure**
## About The Pull Request

Allows for custom regex speech impediments and custom verbs via a tongue augment in the character creation menu. Or at least, that's the idea if I can figure out how to access variable from other files.

## Changelog
:cl:
add: Snowflake tongue
/:cl:

### CODE REVIEW

File: modular_skyrat/modules/customization/modules/client/preferences.dm [hunk 1]
```diff
@@ -2401,6 +2416,32 @@ GLOBAL_LIST_INIT(food, list(
 					ShowLangMenu(user)
 					return TRUE
 
+				if("sverb") //TODO: Get this working
+					var/new_verb = input(user, "Choose your character's speech verb, such as says, hisses, chirps, etc. Do not abuse this feature -- you are expected to stay within the server's HRP guidelines.", "Character Preference")  as text|null
+					if(new_verb)
+						new_verb = reject_bad_name(new_verb) //Stealing this for my own function because fuck that I'm lazy
+						if(!new_verb)
+							to_chat(user, "<font color='red'>Invalid verb. Your verb may only contain the characters A-Z, a-z, - and '.</font>")
+						else if(length(new_verb) < 10 && length(new_verb) > 2)
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/client/preferences.dm:2425]: three seems very short. Struggling to come up with an example of a three letter verb

File: modular_skyrat/modules/customization/modules/client/preferences.dm [hunk 2]
```diff
@@ -2401,6 +2416,32 @@ GLOBAL_LIST_INIT(food, list(
...
+							speech_verb = lowertext(new_verb)
+						else
+							to_chat(user, "<font color='red'>Invalid verb. Your verb must be 3 to 9 characters long. </font>")
+
+				if("speech_rep_from")
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/client/preferences.dm:2430]:
```suggestion
				if("speech_replace_from") 
```

File: modular_skyrat/modules/customization/modules/client/preferences.dm [hunk 3]
```diff
@@ -2401,6 +2416,32 @@ GLOBAL_LIST_INIT(food, list(
...
+
+				if("speech_rep_from") 
+					var/search_char = input(user, "What character(s) do you wish to replace? This supports some basic REGEX.", "Character Preference")  as text|null
+					if(search_char)
+						if(length(search_char) < 6 && length(search_char) >= 0)
+							speech_replace_from = strip_html_simple(search_char, 6, TRUE)
+						else
+							to_chat(user, "<font color='red'>Invalid input. Cannot be more than 5 characters.</font>")
+
+				if("speech_rep_to")
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/client/preferences.dm:2438]:
```suggestion
				if("speech_replace_to") 
```

**Review by tf-4 — CHANGES_REQUESTED** [review]

File: modular_skyrat/modules/customization/modules/client/preferences.dm (hunk 1 above)
Comment (Weblure) [reply modular_skyrat/modules/customization/modules/client/preferences.dm:2425]: 3-9 characters

File: modular_skyrat/modules/customization/modules/client/preferences.dm [hunk 4]
```diff
@@ -960,6 +970,11 @@ GLOBAL_LIST_INIT(food, list(
 										print_style = augment_limb_styles[slot_name]
 									if(chosen_limb.uses_robotic_styles)
 										dat += "<tr style='vertical-align:top'><td width='100%' style='background-color:#16274C'><a href='?_src_=prefs;task=augment_style;slot=[slot_name]'>Style</a>: [print_style]</td></tr>"
+								if(category_name == AUGMENT_CATEGORY_ORGANS && chosen_item)//For Custom Speech Quirks
+									if (chosen_item.name == "Unique tongue")
+										//dat += "<tr style='vertical-align:top'><td width='100%' style='background-color:#96274C'>Custom Speech Quirk<br>Replace <a href='?_src_=prefs;preference=speech_rep_from;task=input'>[speech_replace_from]</a> with <a href='?_src_=prefs;preference=speech_rep_to;task=input'>[speech_replace_to]</a></td></tr>"
```
[the remark this answers is not in the document]
Comment (Weblure) [reply modular_skyrat/modules/customization/modules/client/preferences.dm:975]: my guy, this is just a draft, that comment's there for personal reference/testing
same with the modularizing, that'll come _after_ I get this to even start working
I'll keep these suggestions in mind for the future tho

**Comment by hugmeorelse** [comment]
Regex for every say that happens for a character with a custom speech impediment sounds resource intensive for a game that only allows for a single core to run, assuming that the method deviates from the standard method tg uses. I haven't looked too closely into it.

Also, I can't foresee the need for custom speech impediments, considering there are only so many you can make. I rather have a large preset selection rather than go for a custom regex method if my assumption above holds true. 

I'm leaning heavily towards "no" to this PR if this is the case.

Secondly, this seems like a pain to enforce from a staff standpoint.

**Comment by hugmeorelse** [comment]
Talk to some experienced developers and maintainers and get their feedback if you haven't like our former maintainer, Azarak, for example if they're willing for some feedback on optimization. Try to see if this is a feasible project. While I would love to welcome whatever projects players want to work on, we unfortunately have technical constraints and need to restrict bloat.

**Comment by Weblure** [comment]
> Talk to some experienced developers and maintainers and get their feedback if you haven't like our former maintainer, Azarak, for example if they're willing for some feedback on optimization. Try to see if this is a feasible project. While I would love to welcome whatever projects players want to work on, we unfortunately have technical constraints and need to restrict bloat.

Well, if this isn't acceptable performance-wise, then we'll have to remove the lizard speech as well

**Comment by NullDagaf** [comment]
This is seriously overworked in all the wrong ways. You could have just ported the speech impediment datum from oldbase. You also don't know how to handle static variables at all, and are creating new regex datums every time someone speaks. Regex is incredibly performant when you use it properly - This however, ain't it.

**Comment by Cheshify** [comment]
Just roleplay it out

**Comment by Gandalf2k15** [comment]
**NOT DENYING THIS**

Please repopen once the Feature freeze is done! And the 1557 PR is merged!

**Comment by Weblure** [comment]
> **NOT DENYING THIS**
> 
> Please repopen once the Feature freeze is done! And the 1557 PR is merged!

This is a DRAFT
_reeeeeeee-_

**Comment by Weblure** [comment]
> This is seriously overworked in all the wrong ways. You could have just ported the speech impediment datum from oldbase. You also don't know how to handle static variables at all, and are creating new regex datums every time someone speaks. Regex is incredibly performant when you use it properly - This however, ain't it.

Yeah, I have absolutely no clue what I'm doing, I'm just code-monkeying shit together to see what works
Though the old speech impediment isn't what I'm going for, since this needs to be fully customizable

**Comment by Tastyfish** [comment]
I'd like to point out that the one and only reason to make a draft PR is for code feedback.

**Pull request closed without merging**